Fix pthread_atfork handler bug for multiple device types

diviramon opened the pull request:
Summary:
The original implementation used a static local variable inside a lambda to store the device type for pthread_atfork handlers. Since static local variables are initialized only once in C++, all device types (CUDA, MTIA, XPU, etc.) would share the same value - whichever was registered first. This caused fork handlers for devices other than CUDA to incorrectly set the bad fork flag for CUDA instead of their own device type.

The fix uses a compile-time generated array of template-instantiated fork handler functions, where each device index has its own dedicated handler that correctly sets is_in_bad_fork for that specific device.

We probably did not see this before because it is rare for accelerators to be built together. However, as we already saw in the first iteration of this feature, MTIA and GPU might be built together: https://github.com/pytorch/pytorch/pull/144664#issuecomment-2795189672

Test Plan:
Ran the failing MTIA test that was crashing due to DataLoader workers not detecting bad fork state:
buck2 test //hpc/torchrec/models/examples/gpu_tests:dlrm_mtia_test_device_mode

Verified that the fork handlers now correctly set is_in_bad_fork for both CUDA and MTIA device types in forked child processes.

Reviewed By: klintqinami, patrick-toulme

Differential Revision: D88919988

guangyey reviewed (commented):
Seems duplicated with https://github.com/pytorch/pytorch/pull/166619

diviramon commented:
@pytorchbot label "topic: not user facing"

diviramon committed:
Fix pthread_atfork handler bug for multiple device types (#170188)

Summary:

The original implementation used a static local variable inside a lambda to store the device type for pthread_atfork handlers. Since static local variables are initialized only once in C++, all device types (CUDA, MTIA, XPU, etc.) would share the same value - whichever was registered first. This caused fork handlers for devices other than CUDA to incorrectly set the bad fork flag for CUDA instead of their own device type.

The fix uses a compile-time generated array of template-instantiated fork handler functions, where each device index has its own dedicated handler that correctly sets is_in_bad_fork for that specific device.

We probably did not see this before because it is rare for accelerators to be built together. However, as we already saw in the first iteration of this feature, MTIA and GPU might be built together: https://github.com/pytorch/pytorch/pull/144664#issuecomment-2795189672

Test Plan:
Ran the failing MTIA test that was crashing due to DataLoader workers not detecting bad fork state:
buck2 test //hpc/torchrec/models/examples/gpu_tests:dlrm_mtia_test_device_mode

Verified that the fork handlers now correctly set is_in_bad_fork for both CUDA and MTIA device types in forked child processes.

Reviewed By: klintqinami, patrick-toulme

Differential Revision: D88919988

guangyey reviewed (approved):
LGTM. Personally, I like the simple one. Let's @albanD make the final decision.
